### PR TITLE
Integrated a WSGI (Gunicorn) to Bassa API server 

### PR DESCRIPTION
**Pull request by kmehant**
## Description
Integrated a WSGI to Bassa API flask-socketio server to make it production ready. 
Gunicorn is used to start the server on a gevent worker making it possible to serve a large number of clients.

## Motivation and Context
This change is required to make Bassa API server production ready and increase the capability to serve a large number of clients.

## How Has This Been Tested?
Tested locally.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by kmehant** [comment]
@vivonk @agentmilindu @JaDogg Please review :)

File: components/core/requirements.txt [hunk 1]
```diff
@@ -12,3 +12,4 @@ six==1.10.0
 websocket-client==0.37
 Werkzeug==0.11.4
 SQLAlchemy==1.1.9
+gunicorn==19.9.0
```
Comment (JaDogg) [inline components/core/requirements.txt]: Good

**Comment by kmehant** [comment]
I need suggestions for this. We have an issue using gunicorn with docker and you can find it [here](https://github.com/lyft/confidant/issues/151). What are your suggestions regarding this?

**Comment by JaDogg** [comment]
Last time I tried docker with Gunicorn it worked fine for me. But that was 1 year ago. So If there is a problem now I'm not sure how to solve it 🤔

Gunicorn is a fast WSGI server that can handle connections in multiple processes or workers, we can also use threads in each worker as well. This will allow for a single docker container with backend + gunicorn to process reasonable number of requests.

So this solves a part of the problem that is to make our backend handle lot of requests. Now we need something that can take the full force of the real world such as `nginx` to act as a _reverse proxy_ that will connect to multiple Gunicorn powered backends. So if we do this we are theoretically load balancing at 2 levels.

**Comment by agentmilindu** [comment]
I don't think this is an issue. Docker always binds to `localhost`, only on Windows or OSX one has to worry about the IP, but they are used to that  I guess. Anyway, Bassa is intended to be installed on a server and in most cases they are Linux servers, and in most cases, these people know how to handle the stuff.

**Comment by kmehant** [comment]
> I don't think this is an issue. Docker always binds to `localhost`, only on Windows or OSX one has to worry about the IP, but they are used to that I guess. Anyway, Bassa is intended to be installed on a server and in most cases they are Linux servers, and in most cases, these people know how to handle the stuff.

@agentmilindu @JaDogg @vivonk  bro can you please try this PR branch in your machine, and please let me know if you were able to access the server at port 5000. Thank you 😄

**Comment by agentmilindu** [comment]
I can access. @JaDogg @vivonk  What about you?

**Comment by vivonk** [comment]
> I can access. @JaDogg @vivonk What about you?

Working fine :)

File: components/core/gunicorn.py.ini [hunk 2]
```diff
@@ -0,0 +1,14 @@
+"""gunicorn WSGI server configuration."""
+from multiprocessing import cpu_count
+from os import environ
+
+
+def max_workers():    
+    return cpu_count()
```
Comment (vivonk) [inline components/core/gunicorn.py.ini]: 2 * cpu_count + 1 is the optimized value I think
https://medium.com/building-the-system/gunicorn-3-means-of-concurrency-efbb547674b7

**Comment by vivonk** [comment]
@kmehant with this, if we can have [supervisor](http://supervisord.org/) also then it's going to be a better deployment option. 
https://medium.com/ymedialabs-innovation/deploy-flask-app-with-nginx-using-gunicorn-and-supervisor-d7a93aa07c18

**Comment by kmehant** [comment]
> @kmehant with this, if we can have [supervisor](http://supervisord.org/) also then it's going to be a better deployment option.
> https://medium.com/ymedialabs-innovation/deploy-flask-app-with-nginx-using-gunicorn-and-supervisor-d7a93aa07c18

@vivonk bro, You are absolutely right 😃.  I have incorporated them in this branch, please review!!

File: components/core/gunicorn.py.ini [hunk 3]
```diff
@@ -0,0 +1,14 @@
+"""gunicorn WSGI server configuration."""
```
[the remark this answers is not in the document]
Comment (kmehant) [reply components/core/gunicorn.py.ini]: @JaDogg good catch bro. It should have .py extension.

**Comment by kmehant** [comment]
@vivonk @JaDogg @agentmilindu Please review! 
@vivonk @JaDogg I have addressed the changes, shall we merge?

File: components/core/gunicorn.py.ini (hunk 3 above)
Comment (vivonk) [inline components/core/gunicorn.py.ini]: is `.ini` file still required?

[the remark this answers is not in the document]
Comment (kmehant) [reply components/core/gunicorn.py.ini]: @JaDogg Made the changes 😄

Comment (kmehant) [reply components/core/gunicorn.py.ini]: @vivonk Fixed 😄

**Comment by kmehant** [comment]
@vivonk I have fixed the Codacy errors and squashed the commits. 
Can I have a final review on this PR please 😃

**Pull request merged**